36 value backed storage

konistehrad opened the pull request:
XXX: failing tests, not ready for merge. See comments here: https://github.com/thesecretlab/YarnSpinner/issues/36#issuecomment-202226300

desplesda commented on YarnSpinner/Dialogue.cs:
```diff
@@ -330,54 +349,27 @@ private class StandardLibrary : Library {
-					// Give up and say they're not equal
-					return false;
-
+                    return parameters[0].Equals( parameters[1] );
```
So much nicer. :+1:

desplesda commented on YarnSpinner/Value.cs:
```diff
@@ -0,0 +1,334 @@
+namespace Yarn
+{
+	// A value from inside Yarn.
+	public class Value : IComparable, IComparable<Value> {
```
Breaking this out into its own file is a good idea.

konistehrad replied:
Once the overloaded operators started showing up it was almost mandatory. Commenting all the JS semantics take up a ton of space, haha.

desplesda commented on YarnSpinner/Value.cs:
```diff
@@ -0,0 +1,334 @@
+        
+        // this is primarily a debugging property, allowing
+        // for testing between actual and perceived type
+        internal object rawBacking { get; set; }
```
Do we really need to have this? It means that we're storing data twice, and that makes me nervous. 

If we need a stronger way to ensure that the type that the Value thinks it is is the same as the value that was applied, adding logic to the underlying values (`numberValue`, `stringValue`, etc) that throws an exception if the wrong underlying value was accessed (eg if `numberValue` is accessed, but the Value is a bool, then throw an exception) might be a better idea. What do you think?

konistehrad replied:
I'll come clean here: I added that exclusively as a debug variable initially (to ensure the input to the constructor resulted in good state thruout the rest of the class), then made the mistake of leveraging against it in `Equals` and `GetHashCode`. I'll redo the latter and pull the `rawBacking` prop.

desplesda commented on YarnSpinnerTests/Test.cs:
```diff
@@ -101,13 +102,26 @@ public void TestIndentation()
+			}
+		}
+
+		// [Test()]
```
A lot of the  `Test` attributes have been commented out - were you encountering errors in the unit tests?

konistehrad replied:
Nope, sorry was just reducing the cases surrounding my own test. Was gonna restore them when the final PR dropped. :)

konistehrad commented:
Okay! I updated this, fixed some bad comparisons, updated my test cases, restored the others, and everything looks green across the board! Let me know what you think! :)

desplesda commented:
Noticing that the Travis builds are failing - you can run the tests locally without having to wait for travis by doing this on the terminal:

```
nuget restore YarnSpinner.sln
xbuild /p:Configuration=Release YarnSpinner.sln
mono ./YarnSpinnerTests/bin/Release/YarnSpinnerTests.exe
./run-tests.sh -d Tests/TestCases
````

konistehrad commented:
Just saw that @desplesda sorry. One of those days; tests are passing, you push the code, then tests fail. Egg on my face. 😓

desplesda commented:
Haha, no sweat! Happens to the best of us.

konistehrad commented:
Welp! This one is out of my pay grade: this will exit with 0 on local, but is clearly exiting with 1 on CI. No leads on this, sorry!!

desplesda commented:
Ok, will have a chance to look at this hopefully soon! 👌

konistehrad commented:
It was a failing test case in Options.node, sorry I'm on it.

konistehrad commented:
Adding Type.Undefined and using it introduced some nastiness, despite my best intentions. I checked against Twine, and using an undefined variable, as the initial implementation suggests, should be Type.Null not Type.Undefined. I've pulled references to Type.Undefined and made Type.Null the default.

Extending that, I've made `assert` a variadic function in Twine, allowing a description of the failing test case. While not line numbers or the like (I saw that TODO in there), it does allow for pinpointing the failure state.

Sorry that my onboarding has been a little rocky. I promise I'll get the hang of this test harness. 😉

desplesda commented:
Looking good! I'll play with this a little before merging its but it all looks solid. 

Please bug me if I haven't either merged this or spotted a problem after a few days!

desplesda commented on YarnSpinner/Value.cs:
```diff
@@ -0,0 +1,351 @@
+{
+	// A value from inside Yarn.
+	public class Value : IComparable, IComparable<Value> {
+		public static readonly Value DEFAULT = new Value();
```
This feels a bit weird. It looks like the `DEFAULT` value is only used for the case where a variable cannot be found, and is always null; if that's the case, why isn't it called null? And if that change was made, shouldn't all null values simply be references to this instance?

konistehrad replied:
@desplesda Love it, that fix is in! I wasn't able to remove the default constructor, however, as it's still needed to create Type.Variable Value's, as seen in `Parser.cs:694`. Any guidance you could provide would be much appreciated.

desplesda commented:
Awesome work.
